<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.545">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Daniel Jeffares">
<meta name="dcterms.date" content="2024-01-25">
<title>BIO00066I - BIO00066I Workshop 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">BIO00066I</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./workshop2.html" aria-current="page"> 
<span class="menu-text">Workshop 2</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./workshop3.html"> 
<span class="menu-text">Workshop 3</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./workshop4.html"> 
<span class="menu-text">Workshop 4</span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul class="collapse">
<li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives"><span class="header-section-number">1</span> Learning objectives</a></li>
  <li>
<a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction"><span class="header-section-number">2</span> Introduction</a>
  <ul class="collapse">
<li><a href="#the-biology" id="toc-the-biology" class="nav-link" data-scroll-target="#the-biology"><span class="header-section-number">2.1</span> The biology</a></li>
  <li><a href="#research-questions" id="toc-research-questions" class="nav-link" data-scroll-target="#research-questions"><span class="header-section-number">2.2</span> Research questions</a></li>
  <li><a href="#the-data" id="toc-the-data" class="nav-link" data-scroll-target="#the-data"><span class="header-section-number">2.3</span> The data</a></li>
  </ul>
</li>
  <li>
<a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">3</span> Exercises</a>
  <ul class="collapse">
<li><a href="#getting-started" id="toc-getting-started" class="nav-link" data-scroll-target="#getting-started"><span class="header-section-number">3.1</span> Getting started</a></li>
  <li><a href="#loading-the-data" id="toc-loading-the-data" class="nav-link" data-scroll-target="#loading-the-data"><span class="header-section-number">3.2</span> Loading the data</a></li>
  <li><a href="#exploring-the-data" id="toc-exploring-the-data" class="nav-link" data-scroll-target="#exploring-the-data"><span class="header-section-number">3.3</span> Exploring the data</a></li>
  <li><a href="#cleaning-up" id="toc-cleaning-up" class="nav-link" data-scroll-target="#cleaning-up"><span class="header-section-number">3.4</span> Cleaning up</a></li>
  <li><a href="#save-your-work-regularly" id="toc-save-your-work-regularly" class="nav-link" data-scroll-target="#save-your-work-regularly"><span class="header-section-number">3.5</span> Save your work regularly</a></li>
  <li><a href="#summarise-with-dplyr" id="toc-summarise-with-dplyr" class="nav-link" data-scroll-target="#summarise-with-dplyr"><span class="header-section-number">3.6</span> Summarise with dplyr</a></li>
  <li><a href="#making-plots" id="toc-making-plots" class="nav-link" data-scroll-target="#making-plots"><span class="header-section-number">3.7</span> Making plots</a></li>
  </ul>
</li>
  <li><a href="#next-step-modify-the-code" id="toc-next-step-modify-the-code" class="nav-link" data-scroll-target="#next-step-modify-the-code"><span class="header-section-number">4</span> Next step: modify the code</a></li>
  <li>
<a href="#analysis-with-small-data" id="toc-analysis-with-small-data" class="nav-link" data-scroll-target="#analysis-with-small-data"><span class="header-section-number">5</span> Analysis with small data</a>
  <ul class="collapse">
<li><a href="#plotting-a-small-data-set" id="toc-plotting-a-small-data-set" class="nav-link" data-scroll-target="#plotting-a-small-data-set"><span class="header-section-number">5.1</span> Plotting a small data set</a></li>
  <li><a href="#statistical-tests-with-small-data-and-large-sets" id="toc-statistical-tests-with-small-data-and-large-sets" class="nav-link" data-scroll-target="#statistical-tests-with-small-data-and-large-sets"><span class="header-section-number">5.2</span> Statistical tests with small data and large sets</a></li>
  <li><a href="#comparing-plots-side-by-side" id="toc-comparing-plots-side-by-side" class="nav-link" data-scroll-target="#comparing-plots-side-by-side"><span class="header-section-number">5.3</span> Comparing plots side by side</a></li>
  </ul>
</li>
  <li><a href="#reflection" id="toc-reflection" class="nav-link" data-scroll-target="#reflection"><span class="header-section-number">6</span> Reflection</a></li>
  <li><a href="#the-end" id="toc-the-end" class="nav-link" data-scroll-target="#the-end"><span class="header-section-number">7</span> The end</a></li>
  <li>
<a href="#after-the-workshop" id="toc-after-the-workshop" class="nav-link" data-scroll-target="#after-the-workshop"><span class="header-section-number">8</span> After the workshop</a>
  <ul class="collapse">
<li><a href="#consolidation-exercises" id="toc-consolidation-exercises" class="nav-link" data-scroll-target="#consolidation-exercises"><span class="header-section-number">8.1</span> Consolidation exercises</a></li>
  <li><a href="#planning-for-your-report" id="toc-planning-for-your-report" class="nav-link" data-scroll-target="#planning-for-your-report"><span class="header-section-number">8.2</span> Planning for your report</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">BIO00066I Workshop 2</h1>
<p class="subtitle lead">Cell Biology Data Analysis Workshop 2</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Daniel Jeffares </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 25, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Welcome
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<section id="learning-objectives" class="level1" data-number="1"><h1 data-number="1">
<span class="header-section-number">1</span> Learning objectives</h1>
<p>In this workshop, you will learn <strong><em>technical skills</em></strong> and <strong><em>data science concepts</em></strong>.</p>
<p><strong>Technical skills</strong></p>
<ul>
<li>RStudio skills to manage, view and analyse large data sets</li>
<li>How to create plots in RStudio to assist with exploring and interpreting data</li>
<li>How to make your data analysis reproducible and well-explained</li>
</ul>
<p><strong>Thinking like a data scientist</strong></p>
<ul>
<li>How to interpret data with care</li>
<li>How to integrate knowledge from laboratory work and computational analysis</li>
</ul>
<p>The Module learning outcomes <a href="https://www.york.ac.uk/students/studying/manage/programmes/module-catalogue/module/BIO00066I/latest">can be seen here</a>.</p>
</section><section id="introduction" class="level1" data-number="2"><h1 data-number="2">
<span class="header-section-number">2</span> Introduction</h1>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Philosophy
</div>
</div>
<div class="callout-body-container callout-body">
<p>Workshops are not a test. It’s OK to make mistakes. It’s OK to need help. You should be familiar with independent study content before the workshop, but you don’t need to understand <strong>every detail</strong>. The workshop will hlpe you to build and consolidate your understanding. Tips:</p>
<ul>
<li>there are <strong>no stupid questions</strong> (ask us anything!)</li>
<li>don’t worry about making mistakes (we all make mistakes)</li>
<li>discuss code with your neighbours and with the staff</li>
<li>outside of workshop times, google is a good resource</li>
</ul>
</div>
</div>
<section id="the-biology" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="the-biology">
<span class="header-section-number">2.1</span> The biology</h2>
<p>Today we will look at some data from mesenchymal stem cells (MSCs, sometimes called <em>stromal</em> cells). MCSs are are <em>multipotent</em> cells, than can differentiate into many kinds of mesenchymal cells, including bone, cartilage and fat cells. They are not <em>pluripotent</em>, in that they do have limits about how they differentiate.</p>
<p>MSCs are also:</p>
<ul>
<li>immuno-modulatory (they make cytokines)</li>
<li>heterogeneous (they have subpopulations that are different from each other)</li>
</ul>
<p><strong>The MSCs we are looking at</strong> are from bone marrow femoral head tissues (from the hip bone). These MSCs were <em>immortalised</em> (by transforming them with a telomerase gene), and then cloned by limiting dilution. We will compare the data on cell shape and size from just <strong>two clones</strong>, both obtained from one person’s femoral head tissues. Each <em>clone</em> was derived from a <em>single cell</em> from the person. We call these two clones <em>clone A</em> and <em>clone B</em>. Today, we will use our data analysis skills to explore how these cell line differ. Remember, these two cell lines were both obtained from one person. You probably have these kinds of MSCs too.</p>
</section><section id="research-questions" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="research-questions">
<span class="header-section-number">2.2</span> Research questions</h2>
<ol type="1">
<li>What parameters do we have in the data obtained from the Livecyte machine/method?</li>
<li>Which parameters differ between the clones?</li>
<li>Which parameters are correlated?</li>
</ol></section><section id="the-data" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="the-data">
<span class="header-section-number">2.3</span> The data</h2>
<p>The data we have today are derived from live cell imagining of mesenchymal stem cells with a Livecyte microscope and the software it has. Each clones was cultured and the microscope captured cell shape and size information for thousands of cells from each clone (clone A and clone B). For each clone, we measured three <em>biological replicates</em>.</p>
<p>The Livecyte tracked each cell every 23 minutes, giving each data point a <strong>tracking ID</strong> (which tracks a unique cell ‘object’) and a <strong>lineage ID</strong>, which recalls the lineage of cells as they divide. For example, if cell 1, divided into cell 1A and cell 1B, they are all recorded by the Livecyte as being part of lineage ID 1.</p>
<p>This video shows what Livecyte can do (perhaps turn the sound down, or watch later).</p>
<div class="quarto-video ratio ratio-16x9"><iframe data-external="1" src="https://www.youtube.com/embed/lposS9SNngY?si=XQXYAslkA3vK3Pzh" title="" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen=""></iframe></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Biological replicates and technical replicates
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>A biological replicate</strong> repeats an experiment from different cell types, tissue types, or organisms to see if similar results can be observed.</p>
<p><strong>Technical replicates</strong> are repeat measurements of <em>the same biological material</em>, which help to show how much variation comes from the equipment or different methods (rather from the biology).</p>
<p>Our replicates are almost <em>biological replicates</em>, because we grew the cells three times.</p>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Thinking like a data scientist
</div>
</div>
<div class="callout-body-container callout-body">
<p>Data science can be challenging when you first start. But data science does have core concepts, like any other science.</p>
<ul>
<li>Consider the motivation or scientific question <em>before</em> drawing a plot</li>
<li>Reflect on how the plot or analysis addresses the scientific question</li>
<li>Use each plots to adapt, to inspire new research questions and new inquiries</li>
<li>Ensure that your scripts are reproducible and clearly commented</li>
<li><strong>Consider what the results tell you about the biology</strong></li>
</ul>
</div>
</div>
</section></section><section id="exercises" class="level1" data-number="3"><h1 data-number="3">
<span class="header-section-number">3</span> Exercises</h1>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Develop your R script, carefully.
</div>
</div>
<div class="callout-body-container callout-body">
<p>As we run though the exercises, copy the code to your script. You can cop the code by clicking on the ‘clipboard’ symbol at the top tight of the grey code chunks.</p>
<p><strong>#Comments</strong> In R scripts, lines that start with the hash symbol (<code>#</code>) are comments. Add comments so you can remember what this code does. Commenting code makes it readable, for you and anyone else. Good data science includes clearly commented code.</p>
</div>
</div>
<section id="getting-started" class="level2" data-number="3.1"><h2 data-number="3.1" class="anchored" data-anchor-id="getting-started">
<span class="header-section-number">3.1</span> Getting started</h2>
<ol type="1">
<li>Start RStudio from the Start menu</li>
<li>Open your RStudio project using the dropdown menu at the very <strong>top right</strong> of the RStudio window.</li>
<li>Make a new script then save it with a sensible name that will help you to know later what is in this file. <code>BIO00066I-workshop2.R</code> would work.</li>
<li>Add a comment to the script so you know what it is about, for example</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Data Analysis 2: Cell Biology</span></span>
<span><span class="co">#date 2024-30-01</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="5" type="1">
<li>Clear all the previous data, and load <strong><code>tidyverse</code></strong> package by adding these lines to your script:</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#clear previous data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#load the tidyverse</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="6" type="1">
<li>Finally load another library that allows us to make multi-part plots (this can be useful sometimes). #we need this to make pretty plots with the ‘ggarrange’ package</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#load the ggpubr package for multi-part plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/">ggpubr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="7" type="1">
<li>Make sure all these lines of code are in your script, with comments.</li>
<li>Save the script.</li>
</ol></section><section id="loading-the-data" class="level2" data-number="3.2"><h2 data-number="3.2" class="anchored" data-anchor-id="loading-the-data">
<span class="header-section-number">3.2</span> Loading the data</h2>
<p>In workshop 1, we showed you how to <a href="https://3mmarand.github.io/R4BABS/r4babs1/week-8/workshop.html#importing-data-from-files">import data from files</a>. Today, we load a tab-separated value (TSV) file from a website:</p>
<div class="cell">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://djeffares.github.io/BIO66I/all-cell-data-FFT.filtered.2024-02-22.tsv"</span><span class="op">)</span>,</span>
<span>    col_types <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span><span class="op">(</span></span>
<span>        clone <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        replicate <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        tracking.id<span class="op">=</span><span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        lineage.id<span class="op">=</span><span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
About read_tsv
</div>
</div>
<div class="callout-body-container callout-body">
<p>We use the <code>read_tsv</code> function to read the tab-separated value file. Clicking on the link in the <code>read_tsv</code> takes you to a website about this function. <strong>All the code chunks in these workshops have links like this.</strong> The <code>clone = col_factor(),replicate = col_factor()</code> <em>etc</em> let’s R know that we want these columns to be factors, rather than numeric values. Factors are used to represent categorical data.</p>
</div>
</div>
</section><section id="exploring-the-data" class="level2" data-number="3.3"><h2 data-number="3.3" class="anchored" data-anchor-id="exploring-the-data">
<span class="header-section-number">3.3</span> Exploring the data</h2>
<p>It is important to know what data you have. How many rows and columns <em>etc</em>. There are many ways to do this in R. Here are some of our favourites. Copy and past these into your R script, and try them out.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#look at the data, like an excel table:</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/view.html">view</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#what are the names of the columns?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#how many rows and columns you we have?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#other ways to peek at data:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You may have noticed that we have two clones (cloneA and cloneB). For each clone, we have three replicates. For each replicate, we have many readings of cell widths, cell volume, cell sphericity, and so on.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Which method of do <em>you</em> like the best?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why not take a note of this, and use it all the time?</p>
</div>
</div>
</section><section id="cleaning-up" class="level2" data-number="3.4"><h2 data-number="3.4" class="anchored" data-anchor-id="cleaning-up">
<span class="header-section-number">3.4</span> Cleaning up</h2>
<p>When you ran <code>names(cells)</code> you will have seen <em>many</em> metrics from the Livecyte software. We don’t need all these, and the movement metrics them are unreliable, so let’s remove them. It is simple to remove columns of data, using the <code>select</code> command, like so:</p>
<div class="cell">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#see what we have</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#remove all the movement metrics (which are not reliable)</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">cells</span>, </span>
<span>                <span class="op">-</span><span class="va">position.x</span>,</span>
<span>                <span class="op">-</span><span class="va">position.y</span>, </span>
<span>                <span class="op">-</span><span class="va">pixel.position.x</span>, </span>
<span>                <span class="op">-</span><span class="va">pixel.position.y</span>,</span>
<span>                <span class="op">-</span><span class="va">displacement</span>,</span>
<span>                <span class="op">-</span><span class="va">instantaneous.velocity</span>,</span>
<span>                <span class="op">-</span><span class="va">instantaneous.velocity.x</span>,</span>
<span>                <span class="op">-</span><span class="va">instantaneous.velocity.y</span>,</span>
<span>                <span class="op">-</span><span class="va">track.length</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#check that our data is simpler</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the <code>select</code> function, using the minus sign <code>-</code> before a column name <em>removes it</em>. We can also also use <code>select</code> to define which columns to <em>keep</em>, by omitting the <code>-</code>. For example <code>cells &lt;- select(cells, clone, replicate, volume)</code> will only keep the columns we listed.</p>
</section><section id="save-your-work-regularly" class="level2" data-number="3.5"><h2 data-number="3.5" class="anchored" data-anchor-id="save-your-work-regularly">
<span class="header-section-number">3.5</span> Save your work regularly</h2>
<p><strong>Now save your data and save your script.</strong> This command will save <strong>all</strong> the variables you have loaded, or created so far:</p>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#save all my stuff</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save.image</a></span><span class="op">(</span><span class="st">"BIO00066I-workshop2.Rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can load all the data again with:</p>
<div class="cell">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#load all my stuff from last time</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"BIO00066I-workshop2.Rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="summarise-with-dplyr" class="level2" data-number="3.6"><h2 data-number="3.6" class="anchored" data-anchor-id="summarise-with-dplyr">
<span class="header-section-number">3.6</span> Summarise with dplyr</h2>
<p>The <code>view(cells)</code> command shows that we have many cell shape metrics. We saw a summary of our cell shape metrics with the <code>summary(cells)</code> command above.</p>
<p>We can do even better, with tools from the <code>dplyr</code> package (part of the <code>tidyverse</code>). These allow us to make summaries of this data quickly. These methods will work for <strong>any</strong> data frames of any kind of data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
dplyr is wonderful
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>dplyr</code> is like a set of pliers, helping us to ‘bend’ or ‘reshape’ our data.</p>
</div>
</div>
<p>Here is the command. I will explain it below.</p>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">summary.table</span> <span class="op">&lt;-</span> <span class="va">cells</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">clone</span>, <span class="va">replicate</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span></span>
<span>        volume<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">volume</span><span class="op">)</span>,</span>
<span>        mean.thickness<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">mean.thickness</span><span class="op">)</span>,</span>
<span>        radius<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">radius</span><span class="op">)</span>,</span>
<span>        area<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">area</span><span class="op">)</span>,</span>
<span>        sphericity<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">sphericity</span><span class="op">)</span>,</span>
<span>        length<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">length</span><span class="op">)</span>,</span>
<span>        dry.mass<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">dry.mass</span><span class="op">)</span>,</span>
<span>        length.to.width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">length.to.width</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="what-this-code-does" class="level3" data-number="3.6.1"><h3 data-number="3.6.1" class="anchored" data-anchor-id="what-this-code-does">
<span class="header-section-number">3.6.1</span> What this code does</h3>
<ul>
<li><p>The <code>cells |&gt;</code> part takes the data from the <code>cells</code> data frame, and ‘pipes’ it into the <code>group_by</code> function. The <code>|&gt;</code> symbol as means <strong>put this data into the next bit</strong>.</p></li>
<li><p><code>group_by(clone, replicate)</code> means that make groups of data, according to which clone they are, and which replicate culture they were from.</p></li>
<li><p><code>summarise</code> calculates some summaries of all the data rows (for each clone and replicate). The part <code>volume=median(volume)</code> creates a header called <code>volume</code> and fills this with the median cell volume for each clone and replicate</p></li>
<li><p>Right at the top, <code>summary.table &lt;-</code> stores the results of all the piping in an object called <code>summary.table</code>.</p></li>
</ul>
<p>Have a look at the information we generated with:</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://tibble.tidyverse.org/reference/view.html">view</a></span><span class="op">(</span><span class="va">summary.table</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Do you see any patterns in this data?
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>What does this table tell you about <strong>clone A</strong> and <strong>clone B</strong>?</li>
<li>What does this tell you about mesenchymal stem cells?</li>
</ul>
</div>
</div>
<hr></section></section><section id="making-plots" class="level2" data-number="3.7"><h2 data-number="3.7" class="anchored" data-anchor-id="making-plots">
<span class="header-section-number">3.7</span> Making plots</h2>
<p>It looks like many of the cell shape metrics from above might differ between clones. So let’s look deeper, starting with <code>cell width</code>. We will make a box and whisker plot. We use a small ‘trick’ here: by including <code>fill=replicate</code> we force R to make different plots for each replicate.</p>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>,y<span class="op">=</span><span class="va">width</span>,fill<span class="op">=</span><span class="va">replicate</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="workshop2_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Making plots with ggplot
</div>
</div>
<div class="callout-body-container callout-body">
<p>No matter what shape plot you want, <code>ggplot</code> uses the same syntax.</p>
<ol type="1">
<li>Start by telling R what data you want to use like this: <code>ggplot(some_data_frame,aes(x=x_axis_data, y=y_axis_data) +</code>
</li>
</ol>
<p>In this line, <code>x_axis_data</code> and <code>y_axis_data</code> are columns of some_data_frame.</p>
<ol start="2" type="1">
<li>Then define what the shape plot you want (the ‘geometry’): <code><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot()</a></code>, <code><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram()</a></code> etc.</li>
<li>Add extra things to customise your plot, eg: <code>xlab("my x axis label")</code>
</li>
</ol>
<p>If you are unsure what to do, it’s OK to ask! Googling “ggpplot how to make a boxplot” (or any other plot) will help.</p>
</div>
</div>
<p>It does look like clone A and clone B differ consistently in width, with all repeats. To prove this, we would like to do a statistical test. A Student’s t-Test <code>t.test</code> would work. But t-Test’s assume that the data are normally distributed (like a bell curve).</p>
<p>We can test this approximately by plotting data, as below. It doesn’t look perfectly ‘bell shaped’, so let’s play is safe and use a nonparametric test.</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">width</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The nonparametric equivalent of a t-Test is a <strong>Wilcoxon rank sum test</strong>. There are two ways to run this test in R. We can do <code>wilcox.test(vectorA, vectorB)</code>, where vectorA and vectorB contain the numeric values we want to test.</p>
<p>But does does not suit <code>tidyverse</code> data frames very well. So we will use the <code>wilcox.test(numeric_values ~ category_name, data = some_data_frame)</code> method, like so:</p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Wilcoxon rank sum test</span></span>
<span><span class="co">#To test if cloneA and cloneB have statistically different widths</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">width</span> <span class="op">~</span> <span class="va">clone</span>, data <span class="op">=</span> <span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  width by clone
W = 657964994, p-value &lt; 2.2e-16
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>Because the <code>p-value</code> is <em>very low</em> (&lt; 2.2e-16), this means the null hypothesis (that both sets of numbers come from the same population) is <em>very</em> unlikely to be true.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Non-parametric tests are safe
</div>
</div>
<div class="callout-body-container callout-body">
<p>Parametric tests are based on assumptions about the distribution of the real data. Nonparametric statistics are not based on these assumptions. So they are safer. Many biological metrics need nonparametric tests.</p>
</div>
</div>
<section id="an-easier-way" class="level3" data-number="3.7.1"><h3 data-number="3.7.1" class="anchored" data-anchor-id="an-easier-way">
<span class="header-section-number">3.7.1</span> An easier way</h3>
<p>Fortunately, you can add the results of nonparametric tests to ggplot, by adding <code><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html">stat_compare_means()</a></code> to our plot. Note that below we do not force R to split up the replicates as we did above.</p>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>,y<span class="op">=</span><span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html">stat_compare_means</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section></section><section id="next-step-modify-the-code" class="level1" data-number="4"><h1 data-number="4">
<span class="header-section-number">4</span> Next step: modify the code</h1>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
If there is something you don’t uderstand, it’s OK to ask us!
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>To increase your understanding, run through all the plots in section 3.5 with some other cell shape or size metric. To start, find what metrics are present with:</p>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We suggest using one of these, in place of <code>width</code>:</p>
<ul>
<li>volume</li>
<li>mean.thickness</li>
<li>radius</li>
<li>area</li>
<li>sphericity</li>
<li>length</li>
<li>dry.mass</li>
</ul></section><section id="analysis-with-small-data" class="level1" data-number="5"><h1 data-number="5">
<span class="header-section-number">5</span> Analysis with small data</h1>
<p>So far, we have used <strong>all</strong> the rows in our data frame. But perhaps this isn’t wise, because the Livecyte tracking followed each cell over time. So each measurement of cell shape is not a <em>completely independent</em> measurement. It is more of a technical replicate than a biological replicate.</p>
<p>It is also valuable to learn how to make plots and analyse small data sets, as well as large ones. So not we will create a very small subset of the data, and investigate this.</p>
<p>First, read in the small data set we prepared, and have a quick look, to check what we have.</p>
<div class="cell">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#read in some data from a website</span></span>
<span><span class="va">small.tracking.data</span> <span class="op">&lt;-</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_tsv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span><span class="op">(</span><span class="st">"https://djeffares.github.io/BIO66I/trackingid.small.data.tsv"</span><span class="op">)</span>,</span>
<span>    col_types <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/cols.html">cols</a></span><span class="op">(</span></span>
<span>        clone <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        replicate <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        tracking.id<span class="op">=</span><span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        lineage.id<span class="op">=</span><span class="fu"><a href="https://readr.tidyverse.org/reference/parse_factor.html">col_factor</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#check what we have</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">small.tracking.data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">small.tracking.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="plotting-a-small-data-set" class="level2" data-number="5.1"><h2 data-number="5.1" class="anchored" data-anchor-id="plotting-a-small-data-set">
<span class="header-section-number">5.1</span> Plotting a small data set</h2>
<p>Previously we used <code><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot()</a></code> to compare clone A and clone B. Box and whisker plots (<code>geom_boxplot</code>) and violin plots (<code>geom_violin</code>, which we describe below) are good ways of displaying <em>large</em> amounts of data, but they can be misleading for small data sets. For small data sets it is better to simpler plot all the data points, so the audience has all the information. Let’s compare boxplot and a ggplot method to plot all the data (<code>geom_jitter</code>).</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">small.tracking.data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>, y<span class="op">=</span><span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span>fill<span class="op">=</span><span class="cn">NA</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>,size<span class="op">=</span><span class="fl">3</span>,pch<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">small.tracking.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
if you have time
</div>
</div>
<div class="callout-body-container callout-body">
<p>Try two alternations of this plot.</p>
<ol type="1">
<li><p>Remove the <code>geom_boxplot(fill=NA)+</code> line. Does it display the data more clearly?</p></li>
<li><p>Replace width in <code>y=width</code> with some other cell shape/size parameter. Using <code>names(small.tracking.data)</code> will show what metrics we have.</p></li>
</ol>
</div>
</div>
</section><section id="statistical-tests-with-small-data-and-large-sets" class="level2" data-number="5.2"><h2 data-number="5.2" class="anchored" data-anchor-id="statistical-tests-with-small-data-and-large-sets">
<span class="header-section-number">5.2</span> Statistical tests with small data and large sets</h2>
<p>We use statistical tests like the Wilcoxon rank sum test (<code>wilcox.test</code>) to estimate how likely to be drawn from the same ‘population’ (or distribution). If the P-value is very low (eg: one chance in 10,000 or 0.0001), then we can conclude that the two data sets are indeed different. <strong>But the power of such a test depends on the number of data points.</strong></p>
<p>We can prove that to ourselves by comparing a Wilcox test results from the <code>small.tracking.data</code> and the much larger <code>cells</code> data set:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#test with small.tracking.data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">width</span> <span class="op">~</span> <span class="va">clone</span>, data <span class="op">=</span> <span class="va">small.tracking.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  width by clone
W = 75, p-value = 0.1261
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#test with 'cells' data frame - a *much* larger data set  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">width</span> <span class="op">~</span> <span class="va">clone</span>, data <span class="op">=</span> <span class="va">cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  width by clone
W = 657964994, p-value &lt; 2.2e-16
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
P-values can be misleading
</div>
</div>
<div class="callout-body-container callout-body">
<p>P-values are not the biology. Use them with caution, and your own judgment. Sometimes very small differences, or very weak correlations can be statistically significant, but have little relevance for the biology.</p>
</div>
</div>
</section><section id="comparing-plots-side-by-side" class="level2" data-number="5.3"><h2 data-number="5.3" class="anchored" data-anchor-id="comparing-plots-side-by-side">
<span class="header-section-number">5.3</span> Comparing plots side by side</h2>
<p>Now, lets look plotting the small and large data sets side by side. Sometimes, a plot can give us a better sense of the data that a P-value.</p>
<div class="cell">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#plot with the large data</span></span>
<span><span class="co">#storing the plot in an object called large.data.plot</span></span>
<span><span class="va">large.data.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>, y<span class="op">=</span><span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>,size<span class="op">=</span><span class="fl">3</span>,pch<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html">stat_compare_means</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">200</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"large data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plot with the small data</span></span>
<span><span class="co">#storing the plot in an object called small.data.plot</span></span>
<span><span class="va">small.data.plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">small.tracking.data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>, y<span class="op">=</span><span class="va">width</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>,size<span class="op">=</span><span class="fl">3</span>,pch<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html">stat_compare_means</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">200</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"small data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#create a two panel plot, with large.data.plot and small.data.plot</span></span>
<span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html">ggarrange</a></span><span class="op">(</span><span class="va">large.data.plot</span>,<span class="va">small.data.plot</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>geom_jitter</code> with the large data set looks <em>terrible</em>! This is not a good way to show this data. So remove this part of the plot, and try again. You may also want ot try using <code>geom_violin</code> instead of <code>geom_boxplot</code>.</p>
<p>You will also notice that we set the same <em>y</em> axis range for both plots, with <code>ylim(0,200)</code>. We did this so we can compare the plots side by side.</p>
</section></section><section id="reflection" class="level1" data-number="6"><h1 data-number="6">
<span class="header-section-number">6</span> Reflection</h1>
<ul>
<li>After loading the data, what did we do first?</li>
<li>What did making the plots do that summary statistics (liek average, or median) cannot do?</li>
<li>What did we learn about the two clones?</li>
<li>How do we plot and consider P-values for small and large data sets?</li>
</ul></section><section id="the-end" class="level1" data-number="7"><h1 data-number="7">
<span class="header-section-number">7</span> The end</h1>
<p><strong>This is the end of the workshop for today.</strong> We provide some consolidation exercises, extra material and advice about the R Studio project below.</p>
</section><section id="after-the-workshop" class="level1" data-number="8"><h1 data-number="8">
<span class="header-section-number">8</span> After the workshop</h1>
<p>Look at your script again. Add some comments, so that the next time you look at it, it will make more sense to you. Also, have a go at the consolidation exercises below.</p>
<section id="consolidation-exercises" class="level2" data-number="8.1"><h2 data-number="8.1" class="anchored" data-anchor-id="consolidation-exercises">
<span class="header-section-number">8.1</span> Consolidation exercises</h2>
<section id="violin-plots-geom_violin" class="level3" data-number="8.1.1"><h3 data-number="8.1.1" class="anchored" data-anchor-id="violin-plots-geom_violin">
<span class="header-section-number">8.1.1</span> Violin plots <code>geom_violin</code>
</h3>
<ol type="1">
<li>We used <code><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot()</a></code>to show the differences between categories (clone A and clone B). Try <code>geom_violin</code> instead. Which one do you prefer?</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>,y<span class="op">=</span><span class="va">width</span>,fill<span class="op">=</span><span class="va">replicate</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Now adjust the title with <code>ggtitle(label)</code>, and the axis labels with <code>xlab(label)</code> and <code>ylab(label)</code>, and the <strong>theme</strong> (the general style of the plot), with <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic()</a></code>. There are many themes for ggplot, <a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">which you can see here</a>. Use whatever theme you feel makes the data clear and simple.</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">cells</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">clone</span>,y<span class="op">=</span><span class="va">length</span>,fill<span class="op">=</span><span class="va">replicate</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html">geom_violin</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"put your title here!"</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"my X axis label"</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"my Y axis label"</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="making-the-small-data-set" class="level3" data-number="8.1.2"><h3 data-number="8.1.2" class="anchored" data-anchor-id="making-the-small-data-set">
<span class="header-section-number">8.1.2</span> Making the small data set</h3>
<p>This is how we made the small data set <code>trackingid.small.data</code>. Reading through and/or trying out this code will enhance your skills.</p>
<p>Use the <code>dplyr</code> tools <code>group_by</code> and <code>summarise</code> to calculate the average of all the cell shape measurements for each cell. Each cell is marked by the Livecyte software with a tracking id (<code>tracking.id</code>), so we group cells by this. The code <code>group_by(clone, replicate, tracking.id)</code> will group each measurement by the clone, replicate <strong>and</strong> tracking id, then the <code>summarise</code> part calculates the mean values for all our metrics:</p>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#take the mean of each unique tracking id</span></span>
<span><span class="va">trackingid.summary.table</span><span class="op">&lt;-</span><span class="va">cells</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">clone</span>, <span class="va">replicate</span>, <span class="va">tracking.id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span></span>
<span>        volume<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">volume</span><span class="op">)</span>,</span>
<span>        mean.thickness<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">mean.thickness</span><span class="op">)</span>,</span>
<span>        radius<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">radius</span><span class="op">)</span>,</span>
<span>        area<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">area</span><span class="op">)</span>,</span>
<span>        sphericity<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">sphericity</span><span class="op">)</span>,</span>
<span>        length<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">length</span><span class="op">)</span>,</span>
<span>        width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">width</span><span class="op">)</span>,</span>
<span>        dry.mass<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">dry.mass</span><span class="op">)</span>,</span>
<span>        length.to.width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">length.to.width</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'clone', 'replicate'. You can override
using the `.groups` argument.</code></pre>
</div>
</div>
<p>Then we use the <code>dplyr</code> function <code>sample_n</code>, which samples a number of rows from a data frame at random. If the data frame is grouped (which our data is),the number we set applies to each group:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#collect a random subset, of 5 rows for each tracking id</span></span>
<span><span class="va">trackingid.small.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html">sample_n</a></span><span class="op">(</span><span class="va">trackingid.summary.table</span>, <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#check that we have</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">trackingid.small.data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">trackingid.small.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We write the output to a file:</p>
<div class="cell">
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#output trackingid.small.data as a tab-separated value (tsv) file</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_tsv</a></span><span class="op">(</span><span class="va">trackingid.small.data</span>, file <span class="op">=</span><span class="st">"/Users/dj757/gd/modules/BIO66I/data/trackingid.small.data.tsv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Look at the <code>trackingid.small.data</code> data frame, so we know what we have:</p>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#how many rows do we have?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">trackingid.small.data</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#what does the data look like?</span></span>
<span><span class="fu"><a href="https://tibble.tidyverse.org/reference/view.html">view</a></span><span class="op">(</span><span class="va">trackingid.small.data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="planning-for-your-report" class="level2" data-number="8.2"><h2 data-number="8.2" class="anchored" data-anchor-id="planning-for-your-report">
<span class="header-section-number">8.2</span> Planning for your report</h2>
<p>For this module, 30% of the grade is for an R Studio project. The script that you are creating now is the start of this. Here is some information that will help you to prepare to submit your project.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
For the R Studio project
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a logical folder structure for your analysis. Your submission should include the script you used for your work, any accessory functions and the data itself. Your script should be well-commented, well-organised and follow good practice in use of spacing, indentation and variable naming. It should include all the code required to reproduce data import and formatting as well as the summary information, analyses and figures in your report.</p>
</div>
</div>


</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>